backend/src/apiserver/server/api_converter.go

```diff
@@ -1615,90 +1628,164 @@ func toApiRunDetailV1(r *model.Run) *apiv1beta1.RunDetail {
-			}
-		} else {
-			return nil, util.NewInternalServerError(util.NewInvalidInputError("Run.RunDetails cannot be nil"), "Failed to convert API task to its internal representation")
-		}
-		return toModelTask(apiTaskDetailV2)
-	// TODO(gkcalat): implement runtime details of a task below.
 	case *apiv2beta1.PipelineTaskDetail:
 		apiTaskDetailV2 := t
 		namespace = ""
 		taskId = apiTaskDetailV2.GetTaskId()
 		pipelineName = ""
 		runId = apiTaskDetailV2.GetRunId()
-		executionId = fmt.Sprint(apiTaskDetailV2.GetExecutionId())
+		mlmdExecId = fmt.Sprint(apiTaskDetailV2.GetExecutionId())
 		fingerprint = ""
 		createTime = apiTaskDetailV2.GetCreateTime().GetSeconds()
 		startTime = apiTaskDetailV2.GetStartTime().GetSeconds()
 		finishTime = apiTaskDetailV2.GetEndTime().GetSeconds()
 		name = apiTaskDetailV2.GetDisplayName()
 		parentTaskId = apiTaskDetailV2.GetParentTaskId()
 		state = apiTaskDetailV2.GetState().String()
-		stateHistory = ""
-		inputs = ""
-		outputs = ""
+		if hist, err := toModelRuntimeStatuses(apiTaskDetailV2.GetStateHistory()); err == nil {
+			stateHistory = hist
+		}
+		if inpBytes, err := json.Marshal(apiTaskDetailV2.GetInputs()); err == nil {
+			inputs = string(inpBytes)
+		}
+		if outBytes, err := json.Marshal(apiTaskDetailV2.GetOutputs()); err == nil {
+			outputs = string(outBytes)
+		}
+		for _, c := range apiTaskDetailV2.GetChildTasks() {
+			if c.GetTaskId() != "" {
+				children = append(children, c.GetTaskId())
+			} else {
+				children = append(children, c.GetPodName())
+			}
+		}
+	case workflowapi.NodeStatus:
+		// TODO(gkcalat): parse input and output artifacts
+		wfStatus := t
+		nodeId = wfStatus.ID
+		name = wfStatus.DisplayName
+		state = string(wfStatus.Phase)
+		startTime = wfStatus.StartedAt.Unix()
+		createTime = startTime
+		finishTime = wfStatus.FinishedAt.Unix()
+		children = wfStatus.Children
 	default:
 		return nil, util.NewUnknownApiVersionError("Task", t)
 	}
 	return &model.Task{
 		UUID:              taskId,
+		PodName:           nodeId,
 		Namespace:         namespace,
 		PipelineName:      pipelineName,
 		RunId:             runId,
-		MLMDExecutionID:   executionId,
+		MLMDExecutionID:   mlmdExecId,
 		CreatedTimestamp:  createTime,
 		StartedTimestamp:  startTime,
 		FinishedTimestamp: finishTime,
 		Fingerprint:       fingerprint,
 		Name:              name,
 		ParentTaskId:      parentTaskId,
-		State:             state,
+		State:             model.RuntimeState(state).ToV2(),
 		StateHistory:      stateHistory,
 		MLMDInputs:        inputs,
 		MLMDOutputs:       outputs,
+		ChildrenPods:      children,
 	}, nil
 }
 
+// Converts API tasks details into their internal representations.
+// Supports both v1beta1 and v2beta1 API.
+func toModelTasks(t interface{}) ([]*model.Task, error) {
+	if t == nil {
+		return nil, nil
+	}
+	switch t := t.(type) {
```

Comment:
This method probably has a too generic or flexible interface:
- case []*apiv2beta1.PipelineTaskDetail
- case *apiv2beta1.Run
- case []*apiv1beta1.Task
- case *util.Workflow:
- case workflowapi.Nodes:

Do we need a single method that handles all these types of inputs? 
And the method it calls into, which is `toModelTask`, again handles multiple types at once:
- case *apiv1beta1.Task:
- case *apiv2beta1.PipelineTaskDetail
- case workflowapi.NodeStatus

Reply:
For `toModelTask`, we simply provide interface for converting from v1, v2, and argo types. All three are external to the backend.

For `toModelTasks`:
`[]*apiv2beta1.PipelineTaskDetail`: v2
`*apiv2beta1.Run`: I agree. It is unnecessary. Removed.
`[]*apiv1beta1.Task`: v1
`*util.Workflow`: this is convenient to use as argo's Node do not ave parent `runId`, `Namespace`, and `CreatedTimestamp`. This branch takes care of these fields and wraps around the last case.
`workflowapi.Nodes`: argo

Reply:
`[]*apiv1beta1.Task` is not being used as we don't have task-level status update in v1. I will add it later once I figure out usage of `TaskService` in v1.

Reply:
Merged the `workflowapi.Nodes` and `*util.Workflow`